### ISSUE TITLE
Make `+data` and `+onBeforeRender` cumulative

### ISSUE THREAD

**Issue by brillout** [vikejs/vike#1833]

**Comment by shahyar** [comment]
I like this idea. To be clear, would this mean the following?

1. SSR loads, runs `pages/+data`
2. CSR takes over, reuses data from SSR state of `+data`
3. CSR navigates, still reusing state of `+data`. No `pageContext.json`.
- In the event you have `pages/+data.shared` or `pages/+data.client`, at this point it would still re-run.
4. CSR navigates to `/settings/privacy` which colocates `+Page` with `+data`, which fetches `pageContext.json` via `pages/settings/privacy/+data` (or run it client-side if `.shared|client`)
- `pages/+data` is also run here, and merged with the result of the page-specific `+data`

**Comment by brillout** [comment]
**Edit:** I now disagree with my previous comment — my latest thoughts are summed up at https://github.com/vikejs/vike/labels/topic%3Adata.

<details>
  <summary>Previous comment</summary>
  <br>

No, [the data cache](https://github.com/vikejs/vike/issues/1689#issuecomment-2317018245) isn't meant for [your use case](https://github.com/vikejs/vike/discussions/1822). It's only meant for avoiding re-executing +data collocated to +Layout. Consequently, any +data that is collocated to +Page is never cached.

Actually, I'm realizing that it's ambiguous: Vike cannot tell whether `pages/(some-group)/+data.js` is meant to be 1. a data hook for `pages/(some-group)/+Layout.js` or 2. a data hook that every `pages/(some-group)/**/+Page.js` re-uses.

So I'm thinking the data cache should be disabled by default while the user can opt into it. I guess we can implement the data cache separately then. While we are at it, I wonder whether the cache can be extended for other use cases. I created a feature request for it: https://github.com/vikejs/vike/issues/1841.

</details>

**Comment by dajinchu** [comment]
@brillout Hi, just wondering if you also intend to make `+guard.ts` cumulative?

Example use case: 

A folder of routes should only be accessible to users with certain permissions. We can start with a `+guard.ts` at the root of the folder, but if any sub pages have additional restrictions, we have to re-implement the root `+guard`. 

This is easily solved by extracting permission checks into helper functions, or by adding custom `config` with some RBAC scheme and enforcing access control in a centralized `+onBeforeRender`. 

However, wanted to check if cumulative `+guard` was on the roadmap as that would be another possible solution.

**Comment by brillout** [comment]
@dajinchu Yes, the `+guard` hook is actually part of the plan (`TO-DO/soon/cumulative-hooks`). Eventually, all hooks should be able to be cumulative whenever possible (it doesn't make sense for some hooks).

**Comment by rtritto** [comment]
Perhaps can help, I implemented `+data` cumulative in [vike-lite](https://github.com/node-ecosystem/vike-lite/tree/master), you can see the [doc](https://github.com/node-ecosystem/vike-lite/blob/master/packages/vike-lite/doc/differences/CUMULATIVE_DATA.md) and the [commit](https://github.com/node-ecosystem/vike-lite/commit/e1b7bafd137042e6a90d06d0fd81a694a3efd17c).

**Comment by brillout** [comment]
The main issue is that it's a breaking change, so it's best we do this a bit later.
